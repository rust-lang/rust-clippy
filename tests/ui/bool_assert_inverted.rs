#![warn(clippy::bool_assert_inverted)]

use std::ops::Not;

macro_rules! a {
    () => {
        true
    };
}
macro_rules! b {
    () => {
        true
    };
}

#[derive(Debug, Clone, Copy)]
struct ImplNotTraitWithBool;

impl PartialEq<bool> for ImplNotTraitWithBool {
    fn eq(&self, other: &bool) -> bool {
        false
    }
}

impl Not for ImplNotTraitWithBool {
    type Output = bool;

    fn not(self) -> Self::Output {
        true
    }
}

fn main() {
    let a = ImplNotTraitWithBool;

    assert!(!"a".is_empty());
    assert!("".is_empty());
    assert!(!a);
    assert!(a);

    debug_assert!(!"a".is_empty());
    debug_assert!("".is_empty());
    debug_assert!(!a);
    debug_assert!(a);

    assert!(!"a".is_empty(), "tadam {}", false);
    assert!("".is_empty(), "tadam {}", false);
    assert!(!a, "tadam {}", false);
    assert!(a, "tadam {}", false);

    debug_assert!(!"a".is_empty(), "tadam {}", false);
    debug_assert!("".is_empty(), "tadam {}", false);
    debug_assert!(!a, "tadam {}", false);
    debug_assert!(a, "tadam {}", false);
}
