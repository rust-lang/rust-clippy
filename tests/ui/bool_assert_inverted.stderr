error: used `debug_assert!` with an inverted condition
  --> $DIR/bool_assert_inverted.rs:41:5
   |
LL |     debug_assert!(!"a".is_empty());
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert_eq!(.., false, ..)`
   |
   = note: `-D clippy::bool-assert-inverted` implied by `-D warnings`

error: used `debug_assert!` with an inverted condition
  --> $DIR/bool_assert_inverted.rs:43:5
   |
LL |     debug_assert!(!a);
   |     ^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert_eq!(.., false, ..)`

error: used `debug_assert!` with an inverted condition
  --> $DIR/bool_assert_inverted.rs:51:5
   |
LL |     debug_assert!(!"a".is_empty(), "tadam {}", false);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert_eq!(.., false, ..)`

error: used `debug_assert!` with an inverted condition
  --> $DIR/bool_assert_inverted.rs:53:5
   |
LL |     debug_assert!(!a, "tadam {}", false);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `debug_assert_eq!(.., false, ..)`

error: aborting due to 4 previous errors

